common/src/main/java/io/netty/util/SimpleSizeLimitedMap.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * LRU like lock-free cache implementation.
+ *
+ * It rotates 3 maps: primary, secondary, and empty.
+ */
+final class SimpleSizeLimitedMap<K, V> {
```

Comment:
If the thread-safety isn't a requirement for `wildCardCache` I would instead implement an open-addressing like version with less pointer chasing (similar to https://github.com/real-logic/agrona/blob/master/agrona/src/main/java/org/agrona/collections/IntLruCache.java)

Reply:
As it may be used by multiple EventLoops it needs to be Thread safe

Reply:
If the thread-safety is a strict requirement you can use another approach for a "relaxed"  caching using a fixed size single array of immutable entries: JLS ensure reads/writes to be atomic although won't ensure the order: I've used a similar approach here https://github.com/apache/activemq-artemis/blob/98028cdecc6bd31c86a7d6decfed1961e46be7b2/artemis-commons/src/main/java/org/apache/activemq/artemis/utils/AbstractPool.java#L30.

Reply:
sounds interesting as well... And yep that should be good enough

Reply:
Can you also point me to the relevant JLS doc section?

Reply:
Sure, JLS § 17.5 "The usage model for final fields is a simple one: Set the final fields for an object in that object's constructor; and do not write a reference to the object being constructed in a place where another thread can see it before the object's constructor is finished. If this is followed, then when the object is seen by another thread, that thread will always see the correctly constructed version of that object's final fields. It will also see versions of any object or array referenced by those final fields that are at least as up-to-date as the final fields are."

It's more related to the concept of safe publication: if the object saved in the array is using final fields (that's why I have suggested immutable entry holders) when you reference them using a local variable you're sure they are safely published.

Reply:
Differently we should add a store barrier right after creating a new entry and before writing in the array.

Reply:
@nitsanw should have some experience on that (although he always state he's not a JMM lawyer eheh)

Reply:
@franz1981 thats some nice piece of code and yes it should be good enough for what we do. Considering forking it.

Reply:
Given that I was the original author, but the idea was inspired by http://vanillajava.blogspot.com/2014/12/on-heap-vs-off-heap-memory-usage.html' s `StringInterner`, you can directly apply the whole idea to adapt it for our needs on this and I will be more then happy to review and help on it bud :)